src/property-filter/filtering-token/__tests__/filtering-token.test.tsx

```diff
@@ -1,65 +1,176 @@
+  expect(token.getElement()).toHaveAttribute('role', 'group');
+  expect(token.getElement()).toHaveAccessibleName('filter group with 3 tokens');
+  expect(token.findLabel()).toBe(null);
+  expect(token.findRemoveButton()).toBe(null);
+  expect(token.findTokenOperation()!).toBeNull();
+});
 
-  it('renders a select with and/or text as options', () => {
-    const token = renderToken({ showOperation: true, operation: 'and', andText: 'and-test', orText: 'or-test' });
-    expect(token.findTokenOperation()!.findTrigger().getElement()).toHaveTextContent('and-test');
-    act(() => token.findTokenOperation()!.openDropdown());
-    const dropdown = token.findTokenOperation()!.findDropdown();
-    expect(dropdown.findOptionByValue('and')!.getElement()).toHaveTextContent('and-test');
-    expect(dropdown.findOptionByValue('or')!.getElement()).toHaveTextContent('or-test');
-  });
+test('nested tokens rendered as role="group" with token ARIA label and dismiss button', () => {
+  const token = renderToken({ tokens: [token1, token2, token3], groupAriaLabel: 'filter group with 3 tokens' });
+  const groupTokens = token.findGroupTokens();
+  expect(groupTokens).toHaveLength(3);
 
-  it('fires onChange when a select option is selected', () => {
-    const onChange = jest.fn();
-    const token = renderToken({ showOperation: true, onChange });
-    act(() => token.findTokenOperation()!.openDropdown());
-    act(() => token.findTokenOperation()!.selectOptionByValue('or'));
-    expect(onChange).toHaveBeenCalledWith('or');
-  });
+  for (let index = 0; index < groupTokens.length; index++) {
+    expect(groupTokens[index].getElement()).toHaveAttribute('role', 'group');
+    expect(groupTokens[index].getElement()).toHaveAccessibleName(`filter property${index + 1} = value`);
+    expect(groupTokens[index].findLabel()!.getElement()).toHaveTextContent(`property${index + 1} = value`);
+    expect(groupTokens[index].findRemoveButton()!.getElement()).toHaveAccessibleName(
+      `remove filter property${index + 1} = value`
+    );
+  }
+});
 
-  it('renders the token content', () => {
-    const token = renderToken({ showOperation: true, children: <Popover content="">Hello world</Popover> });
-    expect(token.findLabel().getElement()).toHaveTextContent('Hello world');
-  });
+test('fires root onDismiss when the remove button is pressed', () => {
+  const onDismissToken = jest.fn();
+  const token = renderToken({ tokens: [token1], onDismissToken });
+  token.findRemoveButton().click();
+  expect(onDismissToken).toHaveBeenCalledTimes(1);
+  expect(onDismissToken).toHaveBeenCalledWith(0);
+});
+
+test('fires grouped onDismiss when the remove button is pressed', () => {
+  const onDismissToken = jest.fn();
+  const token = renderToken({ tokens: [token1, token2], onDismissToken });
+  token.findGroupTokens()[1].findRemoveButton().click();
+  expect(onDismissToken).toHaveBeenCalledTimes(1);
```

Comment:
[minor] do we need to check how many times this handler has been called? looks like the statement below is enough for this use case

Reply:
True, but w/o the counter check the test feels incomplete to me - I always add the counter check as a habit.

Say, we might extract the dismiss handler from a lambda into a function and introduce some logic that would handle a dismiss of a single token differently:

```
function handleDismiss(index) {
  if (tokens.length === 1) {
    onDismiss(index);
    // do something else and forget to return
  }
  onDismiss(index);
}
```

In this case the onDimiss will be called twice with the same index which might come unnoticed.